data/transactions/logic/mocktracer/tracer.go

```diff
@@ -200,3 +207,47 @@ func (d *Tracer) AfterOpcode(cx *logic.EvalContext, evalError error) {
 func (d *Tracer) AfterBlock(hdr *bookkeeping.BlockHeader) {
 	d.Events = append(d.Events, AfterBlock(hdr.Round))
 }
+
+// copyDeltas makes a deep copy of the given ledgercore.StateDelta pointer, if it's not nil.
+// This is inefficient, but it should only be used for testing.
+func copyDeltas(deltas *ledgercore.StateDelta) *ledgercore.StateDelta {
+	if deltas == nil {
+		return nil
+	}
+	encoded := protocol.EncodeReflect(deltas)
+	var clone ledgercore.StateDelta
+	err := protocol.DecodeReflect(encoded, &clone)
+	if err != nil {
+		panic(err)
+	}
+	return &clone
+}
+
+// AssertEventsEqual asserts that two slices of Events are equal, taking into account complex
+// equality of StateDeltas. The arguments will be modified in-place to normalize any StateDeltas.
+func AssertEventsEqual(t *testing.T, expected, actual []Event) {
+	t.Helper()
+
+	// Dehydrate deltas for better comparison
+	for i := range expected {
+		if expected[i].Deltas != nil {
+			expected[i].Deltas.Dehydrate()
+		}
+	}
+	for i := range actual {
+		if actual[i].Deltas != nil {
+			actual[i].Deltas.Dehydrate()
+		}
+	}
+
+	// These extra checks are not necessary for correctness, but they provide more targeted information on failure
+	if assert.Equal(t, len(expected), len(actual)) {
+		for i := range expected {
+			jsonExpectedDelta := protocol.EncodeJSONStrict(expected[i].Deltas)
+			jsonActualDelta := protocol.EncodeJSONStrict(actual[i].Deltas)
+			assert.Equal(t, expected[i].Deltas, actual[i].Deltas, "StateDelta disagreement: i=%d, event type: (%v,%v)\n\nexpected: %s\n\nactual: %s", i, expected[i].Type, actual[i].Type, jsonExpectedDelta, jsonActualDelta)
```

Comment:
Could you `assert.Equal` on the json data instead, and thereby get the nice printing of differences?

Reply:
The behavior I'm seeing is that `assert.Equal` will not print out a nice diff with strings. It just prints the entire expected and actual strings, with special characters escaped.

I think that's a bit worse than the current code

Reply:
I agree, I guess the diffs only happen when comparing data structures, not strings. How about `assert.Equal` on maps obtained from:
```
jsonMap := make(map[string](map[string]string))
json.Unmarshal([]byte(b), &jsonMap)
```

Reply:
This is a good idea and I can confirm testify will show a diff when the objects aren't equal.

However, the problem I'm seeing is that this diff is very noisy. (Maybe it's not taking map order into account when generating it?)

For example, if I change just one value in an AccountData, this is what I get from the `require.Equal(t, expected, actual)` statement at the end of the function:

```
Diff:
--- Expected
+++ Actual
@@ -1227,3 +1227,3 @@
			MicroAlgos: (basics.MicroAlgos) {
-         Raw: (uint64) 833333333333333
+         Raw: (uint64) 1666666665665666
			},
```

And this is what I get from comparing the JSON objects as maps:

<details>
<summary>Diff inside (it's much longer)</summary>
<br>

```
Diff:
--- Expected
+++ Actual
@@ -1,2 +1,16 @@
	(map[interface {}]interface {}) (len=9) {
+ (string) (len=10) "Creatables": (map[interface {}]interface {}) (len=2) {
+  (string) (len=4) "1001": (map[interface {}]interface {}) (len=4) {
+   (string) (len=7) "Created": (bool) true,
+   (string) (len=7) "Creator": (string) (len=58) "QT7MZSIZBXL7OHGETGPW6SPKFZPJEUGP3QR5YR4V3EG27IQTP27V4XM2GA",
+   (string) (len=5) "Ctype": (uint64) 1,
+   (string) (len=7) "Ndeltas": (uint64) 0
+  },
+  (string) (len=4) "1003": (map[interface {}]interface {}) (len=4) {
+   (string) (len=7) "Creator": (string) (len=58) "MDUQHKJQF6RHRP5ZXMJVKP2RODF3UM3SBG4L6QSIZP4ZCYAKIMJDAKSLJE",
+   (string) (len=5) "Ctype": (uint64) 1,
+   (string) (len=7) "Ndeltas": (uint64) 0,
+   (string) (len=7) "Created": (bool) true
+  }
+ },
	(string) (len=6) "KvMods": (map[interface {}]interface {}) {
@@ -4,143 +18,158 @@
	(string) (len=13) "PrevTimestamp": (uint64) 0,
+ (string) (len=14) "StateProofNext": (uint64) 0,
	(string) (len=6) "Totals": (map[interface {}]interface {}) {
	},
+ (string) (len=5) "Txids": (map[interface {}]interface {}) (len=2) {
+  (string) (len=44) "k3LdaeBUxaeN6a2pry7t6VijLSn6NAtWF+2w6K9h0Ac=": (map[interface {}]interface {}) (len=2) {
+   (string) (len=5) "Intra": (uint64) 1,
+   (string) (len=9) "LastValid": (uint64) 1001
+  },
+  (string) (len=44) "mlhaDfpAXUQKpPBKnMEx98l7twSN/DbNReIfXYso1wE=": (map[interface {}]interface {}) (len=2) {
+   (string) (len=5) "Intra": (uint64) 0,
+   (string) (len=9) "LastValid": (uint64) 1001
+  }
+ },
+ (string) (len=8) "Txleases": (map[interface {}]interface {}) {
+ },
	(string) (len=5) "Accts": (map[interface {}]interface {}) (len=3) {
+  (string) (len=14) "AssetResources": ([]interface {}) {
+  },
	(string) (len=5) "Accts": ([]interface {}) (len=6) {
	(map[interface {}]interface {}) (len=20) {
+    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
+    (string) (len=6) "Status": (uint64) 0,
+    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
+    },
+    (string) (len=14) "VoteFirstValid": (uint64) 0,
+    (string) (len=13) "VoteLastValid": (uint64) 0,
+    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
+    (string) (len=14) "TotalAppParams": (uint64) 1,
+    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=15) "VoteKeyDilution": (uint64) 0,
+    (string) (len=10) "MicroAlgos": (uint64) 833333333332333,
+    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
+    (string) (len=16) "TotalAssetParams": (uint64) 0,
+    (string) (len=11) "TotalAssets": (uint64) 0,
+    (string) (len=13) "TotalBoxBytes": (uint64) 0,
+    (string) (len=10) "TotalBoxes": (uint64) 0,
		(string) (len=4) "Addr": (string) (len=58) "QT7MZSIZBXL7OHGETGPW6SPKFZPJEUGP3QR5YR4V3EG27IQTP27V4XM2GA",
+    (string) (len=11) "RewardsBase": (uint64) 0,
+    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
+    (string) (len=18) "RewardedMicroAlgos": (uint64) 0
+   },
+   (map[interface {}]interface {}) (len=20) {
+    (string) (len=14) "VoteFirstValid": (uint64) 0,
+    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
+    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
+    (string) (len=6) "Status": (uint64) 2,
+    (string) (len=16) "TotalAssetParams": (uint64) 0,
+    (string) (len=14) "TotalAppParams": (uint64) 0,
+    (string) (len=13) "VoteLastValid": (uint64) 0,
+    (string) (len=11) "TotalAssets": (uint64) 0,
+    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
+    (string) (len=15) "VoteKeyDilution": (uint64) 0,
+    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
+    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
+    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
+    },
+    (string) (len=10) "TotalBoxes": (uint64) 0,
+    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=4) "Addr": (string) (len=58) "FQVGZ2NJU7BIYIUV7UZE655FASFUFQVXVBKIJNUAWHQT2WM35M3BKL5JPY",
+    (string) (len=10) "MicroAlgos": (uint64) 833333333336333,
+    (string) (len=11) "RewardsBase": (uint64) 0,
+    (string) (len=13) "TotalBoxBytes": (uint64) 0
+   },
+   (map[interface {}]interface {}) (len=20) {
+    (string) (len=4) "Addr": (string) (len=58) "374AV2I3UN7NANVBY6A3RYV5JEDUGKRN6VNFXG55457Y3NG3XBYFNX6LNQ",
+    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
+    (string) (len=11) "TotalAssets": (uint64) 0,
+    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
+    (string) (len=13) "VoteLastValid": (uint64) 0,
+    (string) (len=10) "MicroAlgos": (uint64) 0,
+    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=6) "Status": (uint64) 0,
+    (string) (len=13) "TotalBoxBytes": (uint64) 0,
+    (string) (len=15) "VoteKeyDilution": (uint64) 0,
+    (string) (len=11) "RewardsBase": (uint64) 0,
+    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
+    (string) (len=10) "TotalBoxes": (uint64) 0,
+    (string) (len=14) "VoteFirstValid": (uint64) 0,
+    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
+    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
+    (string) (len=14) "TotalAppParams": (uint64) 0,
+    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
+    },
+    (string) (len=16) "TotalAssetParams": (uint64) 0
+   },
+   (map[interface {}]interface {}) (len=20) {
+    (string) (len=4) "Addr": (string) (len=58) "MU54NQW2QTGDKHBKM5V2L3YZDA64C7JW3UFSBK5RQQ6LOWTJY7NWETOGGA",
+    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
+    (string) (len=14) "TotalAppParams": (uint64) 0,
+    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
+    },
+    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
+    (string) (len=15) "VoteKeyDilution": (uint64) 0,
+    (string) (len=10) "MicroAlgos": (uint64) 833333334333333,
+    (string) (len=11) "RewardsBase": (uint64) 0,
+    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
+    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
+    (string) (len=13) "TotalBoxBytes": (uint64) 0,
+    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
+    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=6) "Status": (uint64) 0,
+    (string) (len=16) "TotalAssetParams": (uint64) 0,
+    (string) (len=11) "TotalAssets": (uint64) 0,
+    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=10) "TotalBoxes": (uint64) 0,
+    (string) (len=14) "VoteFirstValid": (uint64) 0,
+    (string) (len=13) "VoteLastValid": (uint64) 0
+   },
+   (map[interface {}]interface {}) (len=20) {
+    (string) (len=14) "VoteFirstValid": (uint64) 0,
+    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
+    (string) (len=16) "TotalAssetParams": (uint64) 0,
+    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
+    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
+    (string) (len=6) "Status": (uint64) 0,
+    (string) (len=13) "VoteLastValid": (uint64) 0,
+    (string) (len=15) "VoteKeyDilution": (uint64) 0,
+    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
+    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
+    (string) (len=14) "TotalAppParams": (uint64) 0,
+    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=11) "TotalAssets": (uint64) 0,
+    (string) (len=13) "TotalBoxBytes": (uint64) 0,
+    (string) (len=10) "TotalBoxes": (uint64) 0,
+    (string) (len=4) "Addr": (string) (len=58) "GJ2SDCPIGHX44WIHGHSSDSB6B64BFSI3LKHHKOBHEYX2GG4RVO2B4466GU",
+    (string) (len=10) "MicroAlgos": (uint64) 1666666665665666,
+    (string) (len=11) "RewardsBase": (uint64) 0,
+    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
+    }
+   },
+   (map[interface {}]interface {}) (len=20) {
+    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
+    (string) (len=16) "TotalAssetParams": (uint64) 0,
+    (string) (len=11) "TotalAssets": (uint64) 0,
+    (string) (len=13) "TotalBoxBytes": (uint64) 0,
+    (string) (len=10) "TotalBoxes": (uint64) 0,
+    (string) (len=4) "Addr": (string) (len=58) "MDUQHKJQF6RHRP5ZXMJVKP2RODF3UM3SBG4L6QSIZP4ZCYAKIMJDAKSLJE",
+    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
+    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
+    (string) (len=14) "TotalAppParams": (uint64) 1,
+    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
+    (string) (len=6) "Status": (uint64) 0,
+    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
+    (string) (len=15) "VoteKeyDilution": (uint64) 0,
+    (string) (len=13) "VoteLastValid": (uint64) 0,
		(string) (len=10) "MicroAlgos": (uint64) 833333333332333,
-    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=13) "VoteLastValid": (uint64) 0,
-    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
-    (string) (len=6) "Status": (uint64) 0,
-    (string) (len=14) "TotalAppParams": (uint64) 1,
-    (string) (len=16) "TotalAssetParams": (uint64) 0,
-    (string) (len=13) "TotalBoxBytes": (uint64) 0,
-    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
-    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
-    (string) (len=11) "RewardsBase": (uint64) 0,
-    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
-    },
-    (string) (len=14) "VoteFirstValid": (uint64) 0,
-    (string) (len=15) "VoteKeyDilution": (uint64) 0,
-    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
-    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
-    (string) (len=11) "TotalAssets": (uint64) 0,
-    (string) (len=10) "TotalBoxes": (uint64) 0
-   },
-   (map[interface {}]interface {}) (len=20) {
-    (string) (len=4) "Addr": (string) (len=58) "FQVGZ2NJU7BIYIUV7UZE655FASFUFQVXVBKIJNUAWHQT2WM35M3BKL5JPY",
-    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
-    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=6) "Status": (uint64) 2,
-    (string) (len=14) "TotalAppParams": (uint64) 0,
-    (string) (len=10) "TotalBoxes": (uint64) 0,
-    (string) (len=13) "VoteLastValid": (uint64) 0,
-    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
-    (string) (len=10) "MicroAlgos": (uint64) 833333333336333,
-    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
-    (string) (len=11) "TotalAssets": (uint64) 0,
-    (string) (len=15) "VoteKeyDilution": (uint64) 0,
-    (string) (len=11) "RewardsBase": (uint64) 0,
-    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
-    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
-    },
-    (string) (len=16) "TotalAssetParams": (uint64) 0,
-    (string) (len=13) "TotalBoxBytes": (uint64) 0,
+    (string) (len=11) "RewardsBase": (uint64) 0,
+    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
+    },
		(string) (len=18) "TotalExtraAppPages": (uint64) 0,
		(string) (len=14) "VoteFirstValid": (uint64) 0
-   },
-   (map[interface {}]interface {}) (len=20) {
-    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=6) "Status": (uint64) 0,
-    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
-    },
-    (string) (len=10) "TotalBoxes": (uint64) 0,
-    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
-    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
-    (string) (len=16) "TotalAssetParams": (uint64) 0,
-    (string) (len=11) "TotalAssets": (uint64) 0,
-    (string) (len=10) "MicroAlgos": (uint64) 0,
-    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
-    (string) (len=11) "RewardsBase": (uint64) 0,
-    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
-    (string) (len=13) "VoteLastValid": (uint64) 0,
-    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
-    (string) (len=14) "VoteFirstValid": (uint64) 0,
-    (string) (len=15) "VoteKeyDilution": (uint64) 0,
-    (string) (len=4) "Addr": (string) (len=58) "374AV2I3UN7NANVBY6A3RYV5JEDUGKRN6VNFXG55457Y3NG3XBYFNX6LNQ",
-    (string) (len=14) "TotalAppParams": (uint64) 0,
-    (string) (len=13) "TotalBoxBytes": (uint64) 0
-   },
-   (map[interface {}]interface {}) (len=20) {
-    (string) (len=11) "RewardsBase": (uint64) 0,
-    (string) (len=6) "Status": (uint64) 0,
-    (string) (len=13) "TotalBoxBytes": (uint64) 0,
-    (string) (len=10) "TotalBoxes": (uint64) 0,
-    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
-    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
-    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
-    },
-    (string) (len=16) "TotalAssetParams": (uint64) 0,
-    (string) (len=14) "VoteFirstValid": (uint64) 0,
-    (string) (len=15) "VoteKeyDilution": (uint64) 0,
-    (string) (len=13) "VoteLastValid": (uint64) 0,
-    (string) (len=4) "Addr": (string) (len=58) "MU54NQW2QTGDKHBKM5V2L3YZDA64C7JW3UFSBK5RQQ6LOWTJY7NWETOGGA",
-    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
-    (string) (len=14) "TotalAppParams": (uint64) 0,
-    (string) (len=10) "MicroAlgos": (uint64) 833333334333333,
-    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
-    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
-    (string) (len=11) "TotalAssets": (uint64) 0
-   },
-   (map[interface {}]interface {}) (len=20) {
-    (string) (len=4) "Addr": (string) (len=58) "GJ2SDCPIGHX44WIHGHSSDSB6B64BFSI3LKHHKOBHEYX2GG4RVO2B4466GU",
-    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
-    (string) (len=11) "RewardsBase": (uint64) 0,
-    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
-    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=10) "MicroAlgos": (uint64) 833333333333333,
-    (string) (len=6) "Status": (uint64) 0,
-    (string) (len=16) "TotalAssetParams": (uint64) 0,
-    (string) (len=10) "TotalBoxes": (uint64) 0,
-    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
-    (string) (len=15) "VoteKeyDilution": (uint64) 0,
-    (string) (len=13) "TotalBoxBytes": (uint64) 0,
-    (string) (len=14) "VoteFirstValid": (uint64) 0,
-    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
-    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
-    (string) (len=14) "TotalAppParams": (uint64) 0,
-    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
-    },
-    (string) (len=11) "TotalAssets": (uint64) 0,
-    (string) (len=13) "VoteLastValid": (uint64) 0
-   },
-   (map[interface {}]interface {}) (len=20) {
-    (string) (len=4) "Addr": (string) (len=58) "MDUQHKJQF6RHRP5ZXMJVKP2RODF3UM3SBG4L6QSIZP4ZCYAKIMJDAKSLJE",
-    (string) (len=19) "TotalAppLocalStates": (uint64) 0,
-    (string) (len=11) "TotalAssets": (uint64) 0,
-    (string) (len=13) "TotalBoxBytes": (uint64) 0,
-    (string) (len=10) "TotalBoxes": (uint64) 0,
-    (string) (len=6) "VoteID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=11) "SelectionID": (string) (len=44) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
-    (string) (len=6) "Status": (uint64) 0,
-    (string) (len=18) "TotalExtraAppPages": (uint64) 0,
-    (string) (len=14) "VoteFirstValid": (uint64) 0,
-    (string) (len=10) "MicroAlgos": (uint64) 833333333332333,
-    (string) (len=18) "RewardedMicroAlgos": (uint64) 0,
-    (string) (len=12) "StateProofID": (string) (len=88) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
-    (string) (len=14) "TotalAppSchema": (map[interface {}]interface {}) {
-    },
-    (string) (len=16) "TotalAssetParams": (uint64) 0,
-    (string) (len=15) "VoteKeyDilution": (uint64) 0,
-    (string) (len=8) "AuthAddr": (string) (len=58) "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
-    (string) (len=11) "RewardsBase": (uint64) 0,
-    (string) (len=14) "TotalAppParams": (uint64) 1,
-    (string) (len=13) "VoteLastValid": (uint64) 0
	}
@@ -159,4 +188,4 @@
		(string) (len=5) "State": (map[interface {}]interface {}) (len=2) {
-     (string) (len=7) "Deleted": (bool) false,
-     (string) (len=10) "LocalState": (interface {}) <nil>
+     (string) (len=10) "LocalState": (interface {}) <nil>,
+     (string) (len=7) "Deleted": (bool) false
		}
@@ -178,46 +207,17 @@
	}
-  },
-  (string) (len=14) "AssetResources": ([]interface {}) {
	}
	},
- (string) (len=10) "Creatables": (map[interface {}]interface {}) (len=2) {
-  (string) (len=4) "1001": (map[interface {}]interface {}) (len=4) {
-   (string) (len=7) "Created": (bool) true,
-   (string) (len=7) "Creator": (string) (len=58) "QT7MZSIZBXL7OHGETGPW6SPKFZPJEUGP3QR5YR4V3EG27IQTP27V4XM2GA",
-   (string) (len=5) "Ctype": (uint64) 1,
-   (string) (len=7) "Ndeltas": (uint64) 0
-  },
-  (string) (len=4) "1003": (map[interface {}]interface {}) (len=4) {
-   (string) (len=7) "Created": (bool) true,
-   (string) (len=7) "Creator": (string) (len=58) "MDUQHKJQF6RHRP5ZXMJVKP2RODF3UM3SBG4L6QSIZP4ZCYAKIMJDAKSLJE",
-   (string) (len=5) "Ctype": (uint64) 1,
-   (string) (len=7) "Ndeltas": (uint64) 0
-  }
- },
	(string) (len=3) "Hdr": (map[interface {}]interface {}) (len=12) {
+  (string) (len=4) "rate": (uint64) 1666666666,
+  (string) (len=3) "rnd": (uint64) 1,
+  (string) (len=6) "rwcalr": (uint64) 500000,
+  (string) (len=3) "rwd": (string) (len=58) "7777777777777777777777777777777777777777777777777774MSJUVU",
	(string) (len=2) "ts": (uint64) 1683911424,
-  (string) (len=6) "txn256": (string) (len=44) "HQn2+iMjWIFRTaDl+PeJdn5rKOch+otqMK5/Gao+puc=",
	(string) (len=4) "frac": (uint64) 1666666666,
+  (string) (len=3) "gen": (string) (len=4) "test",
+  (string) (len=2) "gh": (string) (len=44) "Ed0uTVbpak8+nKNd5xXbbVQpcvLOyjkLgQlrBGf8c+4=",
	(string) (len=4) "prev": (string) (len=56) "blk-5H7NNKRELCPEIY4WOSHLSV2D2OORB7US4OS75ZSSAIIPCX7V5EDQ",
	(string) (len=5) "proto": (string) (len=6) "future",
-  (string) (len=3) "rnd": (uint64) 1,
-  (string) (len=6) "rwcalr": (uint64) 500000,
-  (string) (len=4) "fees": (string) (len=58) "FQVGZ2NJU7BIYIUV7UZE655FASFUFQVXVBKIJNUAWHQT2WM35M3BKL5JPY",
-  (string) (len=3) "gen": (string) (len=4) "test",
-  (string) (len=2) "gh": (string) (len=44) "Ed0uTVbpak8+nKNd5xXbbVQpcvLOyjkLgQlrBGf8c+4=",
-  (string) (len=4) "rate": (uint64) 1666666666,
-  (string) (len=3) "rwd": (string) (len=58) "7777777777777777777777777777777777777777777777777774MSJUVU"
- },
- (string) (len=14) "StateProofNext": (uint64) 0,
- (string) (len=5) "Txids": (map[interface {}]interface {}) (len=2) {
-  (string) (len=44) "k3LdaeBUxaeN6a2pry7t6VijLSn6NAtWF+2w6K9h0Ac=": (map[interface {}]interface {}) (len=2) {
-   (string) (len=5) "Intra": (uint64) 1,
-   (string) (len=9) "LastValid": (uint64) 1001
-  },
-  (string) (len=44) "mlhaDfpAXUQKpPBKnMEx98l7twSN/DbNReIfXYso1wE=": (map[interface {}]interface {}) (len=2) {
-   (string) (len=5) "Intra": (uint64) 0,
-   (string) (len=9) "LastValid": (uint64) 1001
-  }
- },
- (string) (len=8) "Txleases": (map[interface {}]interface {}) {
+  (string) (len=6) "txn256": (string) (len=44) "HQn2+iMjWIFRTaDl+PeJdn5rKOch+otqMK5/Gao+puc=",
+  (string) (len=4) "fees": (string) (len=58) "FQVGZ2NJU7BIYIUV7UZE655FASFUFQVXVBKIJNUAWHQT2WM35M3BKL5JPY"
	}
```

</details>

Reply:
I think I misunderstood why you're doing the json conversion. I thought the normal message provided by `require.Equals` was too noisy, so you converted to json so that it would be easier for a human to compare.
But I feel like you're telling me the diff on the Delta is fine.  So what is the json for?

Reply:
The JSON printing was useful during development because of two reasons:
1. There were a lot of problems at the beginning, and attempting to find out which differences belong to which StateDeltas was a challenge because the diff was so large.
2. Testify will stop showing diffs at a certain depth, so some information was hidden.

But just now I've edited the expected objects in a few different ways and the diffs produced by testify seem pretty good. I think I'll remove the JSON stuff and only keep a delta-by-delta comparison so you see which delta caused a failure.

Reply:
Sounds good, I definitely see the utility in getting the index specific comparison.